### PR TITLE
Redhat FIPS selftest

### PR DESCRIPTION
**Pull request by beldmit**
This PR brings a set of Red Hat KATs FIPS-related patches upstream.

Original patches were implemented mostly by @neverpanic

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by beldmit** [comment]
**OTC: which patches of this subset are desired?**

I am going to fix the tests after we agree on which is acceptable.

File: crypto/evp/m_sigver.c [hunk 1]
```diff
@@ -23,6 +23,7 @@ static int update(EVP_MD_CTX *ctx, const void *data, size_t datalen)
     ERR_raise(ERR_LIB_EVP, EVP_R_ONLY_ONESHOT_SUPPORTED);
     return 0;
 }
+#endif /* !defined(FIPS_MODULE) */
```
Comment (paulidale) [inline crypto/evp/m_sigver.c:26]: Why the changes in this file?  These would apply to the FIPS provider's internal libctx and it doesn't need access to these when performing other operations (which is the only reason it includes the EVP bindings).

File: providers/fips/self_test_data.inc [hunk 2]
```diff
@@ -824,188 +824,190 @@ static const ST_KAT_DRBG st_kat_drbg_tests[] =
 
 #ifndef OPENSSL_NO_DH
 /* DH KAT */
+/* RFC7919 FFDHE2048 p */
```
Comment (paulidale) [inline providers/fips/self_test_data.inc:827]: Why is this change required?
What's wrong with the existing self test?

Comment (neverpanic) [reply providers/fips/self_test_data.inc:827]: [NIST SP 800-56Ar3](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf) says in section 5.5.1.1 on DH params:

> Two classes of domain parameters are approved for FFC key agreement: a class of “safe” domain parameters that are associated with approved safe-prime groups, and a class of “FIPS 186-type” domain parameters that conform to one of the FIPS 186-type parameter-size sets that are listed in Table 1.

A footnote on the FIPS 186-type domain parameters points to "FIPS 186, Digital Signature Standard (DSS)". [FIPS 186-4](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf) does specify those domain params in section 4 "DSA" subsection 4.3.1 "Domain Parameter Generation" and Appendix A. However, the newer version [FIPS 186-5](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-5.pdf) no longer specifies DSA at all; it just says:

> Prior versions of this standard specified the DSA. This standard no longer approves the DSA for digital signature generation. However, the DSA may be used to verify signatures generated prior to the implementation date of this standard. See FIPS 186-4 for the specifications for the DSA.

You don't need generation of FIPS 186-type domain parameters for signature verification.

We've been told by our lab that once the transition period for DSA ends, submissions will no longer be able to claim DSA, and with that FIPS 186-type domain parameters will automatically become non-approved as well. I believe the transition period has not yet started, but when it does, use of a FIPS 186-type group in the self-test would become unapproved after a year. For this reason, we swapped it with a safe-prime group and suggest you do the same to be future-proof.

Comment (paulidale) [reply providers/fips/self_test_data.inc:827]: Thanks for the explanation.

**Comment by beldmit** [comment]
> Use OAEP in FIPS KAT tests

No review yet.

> Use combination of DigestSign/DigestVerify functions in FIPS KATs

No review yet, the explanation is provided

> Use well-known group FFDHE2048 in self-test

The rationale is provided

**Comment by paulidale** [comment]
* OAEP: I'd prefer to see a test RNG setup, rather than hacking in a seed.
* DigestSign/Verify: I could live with this but the preprocessor mess in m_sigver.h might be excessive.
* FFDHE2048: I'm okay with this change.

**Comment by neverpanic** [comment]
> DigestSign/Verify: I could live with this but the preprocessor mess in m_sigver.h might be excessive.

That is indeed unfortunate, and I had the same thought while writing it.
Perhaps an alternative would be to use the provider functions directly instead, but that's also not necessarily simpler to implement, either.

**Comment by paulidale** [comment]
It would be possible to leave the legacy paths in, they'd not be used by the FIPS provider so should be harmless.
It's ugly, but in a different way.

**Comment by neverpanic** [comment]
The reason I removed the legacy paths using `#ifdef`s is that not doing so required a seemingly endless number of additional functions that were not yet compiled into the FIPS provider, which in turn required many more additional `#ifdef`s (or removal of existing ones) in other places.

File: crypto/evp/m_sigver.c [hunk 3]
```diff
@@ -469,23 +479,29 @@ int EVP_DigestVerifyUpdate(EVP_MD_CTX *ctx, const void *data, size_t dsize)
     return EVP_DigestUpdate(ctx, data, dsize);
 }
 
-#ifndef FIPS_MODULE
+
```
Comment (levitte) [inline crypto/evp/m_sigver.c:482]: Nit: Remove the extra empty line

File: crypto/rsa/rsa_oaep.c [hunk 4]
```diff
@@ -44,19 +44,24 @@ int RSA_padding_add_PKCS1_OAEP(unsigned char *to, int tlen,
                                                    param, plen, NULL, NULL);
 }
 
+#ifdef FIPS_MODULE
+extern int REDHAT_FIPS_asym_cipher_st;
+#endif /* FIPS_MODULE */
```
Comment (levitte) [inline crypto/rsa/rsa_oaep.c:49]: ?

File: crypto/rsa/rsa_oaep.c [hunk 5]
```diff
@@ -107,6 +112,11 @@ int ossl_rsa_padding_add_PKCS1_OAEP_mgf1_ex(OSSL_LIB_CTX *libctx,
     db[emlen - flen - mdlen - 1] = 0x01;
     memcpy(db + emlen - flen - mdlen, from, (unsigned int)flen);
     /* step 3d: generate random byte string */
+#ifdef FIPS_MODULE
+    if (redhat_st_seed != NULL && REDHAT_FIPS_asym_cipher_st) {
+        memcpy(seed, redhat_st_seed, mdlen);
+    } else
+#endif
```
Comment (levitte) [inline crypto/rsa/rsa_oaep.c:119]: ?

File: providers/fips/self_test_kats.c [hunk 6]
```diff
@@ -641,14 +650,21 @@ static int self_test_ciphers(OSSL_SELF_TEST *st, OSSL_LIB_CTX *libctx)
     return ret;
 }
 
+int REDHAT_FIPS_asym_cipher_st = 0;
```
Comment (levitte) [inline providers/fips/self_test_kats.c:653]: ?

File: providers/fips/self_test_kats.c [hunk 7]
```diff
@@ -641,14 +650,21 @@ static int self_test_ciphers(OSSL_SELF_TEST *st, OSSL_LIB_CTX *libctx)
...
+
 static int self_test_asym_ciphers(OSSL_SELF_TEST *st, OSSL_LIB_CTX *libctx)
 {
     int i, ret = 1;
 
+    REDHAT_FIPS_asym_cipher_st = 1;
```
Comment (levitte) [inline providers/fips/self_test_kats.c:659]: ?

File: providers/fips/self_test_kats.c [hunk 8]
```diff
@@ -641,14 +650,21 @@ static int self_test_ciphers(OSSL_SELF_TEST *st, OSSL_LIB_CTX *libctx)
...
+
     for (i = 0; i < (int)OSSL_NELEM(st_kat_asym_cipher_tests); ++i) {
         if (!self_test_asym_cipher(&st_kat_asym_cipher_tests[i], st, libctx))
             ret = 0;
     }
+
+    REDHAT_FIPS_asym_cipher_st = 0;
```
Comment (levitte) [inline providers/fips/self_test_kats.c:666]: ?

File: providers/implementations/asymciphers/rsa_enc.c [hunk 9]
```diff
@@ -430,6 +451,10 @@ static const OSSL_PARAM *rsa_gettable_ctx_params(ossl_unused void *vprsactx,
     return known_gettable_ctx_params;
 }
 
+#ifdef FIPS_MODULE
+extern int REDHAT_FIPS_asym_cipher_st;
```
Comment (levitte) [inline providers/implementations/asymciphers/rsa_enc.c:455]: ?

File: providers/implementations/asymciphers/rsa_enc.c [hunk 10]
```diff
@@ -541,6 +566,18 @@ static int rsa_set_ctx_params(void *vprsactx, const OSSL_PARAM params[])
         prsactx->oaep_labellen = tmp_labellen;
     }
 
+#ifdef FIPS_MODULE
+    p = OSSL_PARAM_locate_const(params, OSSL_ASYM_CIPHER_PARAM_REDHAT_KAT_OEAP_SEED);
+    if (p != NULL && REDHAT_FIPS_asym_cipher_st) {
```
Comment (levitte) [inline providers/implementations/asymciphers/rsa_enc.c:571]: ?

**Review by levitte — CHANGES_REQUESTED** [review]
A name with "REDHAT" in it has me raise an eyebrow sharply.  I hope this isn't really something RedHat specific, in which case a rename would be a good thing.  If this *is* something RedHat specific, it needs some rethinking to be generalised.

**Comment by beldmit** [comment]
Quoting the OTC discussion:

> No concerns in principle but there are review issues to be resolved.
